src/dune_engine/load_rules.ml

```diff
@@ -501,14 +501,15 @@ end = struct
 
   module Normal = struct
     type t =
-      { build_dir_only_sub_dirs : Subdir_set.t
+      { build_dir_only_sub_dirs : Subdir_set.t Path.Build.Map.t
       ; directory_targets : Loc.t Path.Build.Map.t
       ; rules : Rules.t Memo.Lazy.t
       }
 
     let combine_exn r { build_dir_only_sub_dirs; directory_targets; rules } =
       { build_dir_only_sub_dirs =
-          Subdir_set.union r.build_dir_only_sub_dirs build_dir_only_sub_dirs
+          Path.Build.Map.union r.build_dir_only_sub_dirs build_dir_only_sub_dirs
+            ~f:(fun _ a b -> Some (Subdir_set.union a b))
```

Comment:
Why is this code written twice? Please factor it out.

Reply:
Thanks, added a new module in [522eaea](https://github.com/ocaml/dune/pull/7207/commits/522eaead7b8b3c01d3a67573fc0d4fb346a568be) with this logic. I think it becomes more explicit.

Reply:
Thanks! Agreed that it's better now.